### PR TITLE
Add a function that constructs samplers

### PR DESCRIPTION
**Pull request by brandonwillard**
This PR provides an initial implementation of #3&mdash;and the related interface function mentioned in #26.  It provides a general sampler-constructor, `aemcmc.basic.construct_sampler`, that returns a `dict` mapping `RandomVariable`s to their sample steps.

The current approach uses a `Feature` called `SamplerTracker` to track a `dict` from `RandomVariable`s to all their discovered sample steps&mdash;even when there's more than one potential sampler for the same `RandomVariable`.  Sample steps are discovered by walking the graph with standard local rewriters that write their results to the `dict` in `SamplerTracker`.  This allows us to maintain the original observation variable graphs in relation to every other un-observed variable (i.e. so we can see when a variable is in a particular hierarchical relationship with another variable, etc.)

In order to get around some `DimShuffle` annoyances during unification/pattern-matching, a `SubsumingElemwise` `Op` was added and is used to replace `Elemwise(DimShuffle(x), ...)` graphs with `SubsumingElemwise(x, ...)` graphs (i.e. ones that subsume the `DimShuffle`s).  Since `SubsumingElemwise` inherits from `OpFromGraph`, those nodes can be expanded later on to reproduce the original `Elemwise` + `DimShuffle` sub-graphs.

- [x] Work out the design of the main "processing loop"
    This loop needs to iterate over all the unobserved `RandomVariable`s and construct samplers for them.  While doing so, it needs to use references to the previously constructed samplers' outputs.
- [x] Do something about `RandomVariable` rewriting
    The issue here is that `RandomVariable`s that are canonicalized are no longer the same `RandomVariable`s that the user created, so we need a means of keeping a map between the two.  Lifting `Op`s through `RandomVariable`s is one of the main ways this issue shows up.  N.B.: This is also one situation in which we could use complete relations (i.e. two-way rewrites).
- [x] Restrict the kinds of `DimShuffle`s `SubsumingElemwise` will subsume (i.e. limit to only ones that add the appropriate broadcast dimensions).
    In cases where the original graph was a `Elemwise(DimShuffle1(DimShuffle2(x)), ...)` and the two `DimShuffle*` are merged, we will need to un-merge/expand them in order to use `SubsumingElemwise`.
- [x] Add initial value checks to tests (i.e. make sure that initial value variables correctly replace their corresponding `RandomVariable`s).
- [x] Finish refactoring `aemcmc.gibbs` (e.g. remove combination samplers in favor of `construct_sampler`, create more `local_optimizer`s, generalize `local_optimizer` construction, etc.)

### CODE REVIEW

File: aemcmc/conjugates.py [hunk 1]
```diff
@@ -1,12 +1,10 @@
 import aesara.tensor as at
+from aeppl.opt import NoCallbackEquilibriumDB
```
Comment (rlouf) [inline aemcmc/conjugates.py]: If `NoCallbackEquilibriumDB` were to change in `aeppl` would we necessarily want this change to be propagated in `aemcmc`? Wondering if it make more sense to subclass `EquilibriumDB` directly in `opt.py` here.

File: aemcmc/gibbs.py [hunk 2]
```diff
@@ -86,26 +95,16 @@ def update_beta(
...
-            Biometrika, 97(2), 465-480.
-
-    """
-    size = at.scalar("size", dtype="int32")
-    tau_rv = srng.halfcauchy(0, 1, size=1)
-    lmbda_rv = srng.halfcauchy(0, 1, size=size)
-    beta_rv = srng.normal(0, tau_rv * lmbda_rv, size=size)
-    return beta_rv
-
+def horseshoe_match(
```
[the remark this answers is not in the document]
Comment (brandonwillard) [reply aemcmc/gibbs.py:98]: Yeah, all of the `aemcmc.gibbs` code is going to be completely refactored to use `construct_sampler`, so that will go away.

**Comment by brandonwillard** [comment]
We now have a complete working example on the test `test_basic.py:test_create_gibbs`.

**Comment by brandonwillard** [comment]
I've finished refactoring the sampler steps and filled out the docstings, so this should be ready to merge when/if it passes.

**Pull request merged**